File: backend/app/schemas/campaign.py
"""Pydantic v2 schemas for Campaign and CampaignRecipient CRUD."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import CampaignStatus, RecipientStatus

# ─── Campaign schemas ───────────────────────────────────────────────


class CampaignCreate(BaseModel):
    """Schema for creating a new campaign.

    audience_filter must contain at least one criterion (e.g. tags, language).
    Contacts with opt_in_status = opted_out are excluded at the service layer.
    """

    name: str = Field(..., min_length=1, max_length=200, description="Campaign name")
    description: str | None = Field(default=None, description="Optional description")
    template_id: str = Field(..., description="Meta template ID")
    template_name: str = Field(..., description="Human-readable template name")
    template_language: str = Field(default="fr", description="Template language code")
    audience_filter: dict = Field(
        ..., description="Targeting criteria: tags, segments, language, etc."
    )
    variable_mapping: dict = Field(
        default_factory=dict, description='Template variable mapping: {"1": "contact.name", ...}'
    )

    @model_validator(mode="after")
    def validate_audience_filter(self) -> CampaignCreate:
        """Vérifie que audience_filter contient au moins un critère."""
        if not self.audience_filter:
            raise ValueError("audience_filter must contain at least one criterion")
        return self


class CampaignUpdate(BaseModel):
    """Schema for updating a campaign. All fields optional.

    Only editable when campaign status is 'draft' (enforced at service layer).
    """

    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    audience_filter: dict | None = None
    variable_mapping: dict | None = None


class CampaignRead(BaseModel):
    """Full campaign response — returned by the API."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str | None
    template_id: str
    template_name: str
    template_language: str
    audience_filter: dict
    audience_count: int
    variable_mapping: dict
    status: CampaignStatus
    scheduled_at: datetime | None
    started_at: datetime | None
    completed_at: datetime | None
    stats: dict
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime


class CampaignList(BaseModel):
    """Paginated list of campaigns."""

    items: list[CampaignRead]
    total: int
    page: int
    page_size: int


class CampaignSchedule(BaseModel):
    """Schema for scheduling a campaign send."""

    scheduled_at: datetime = Field(..., description="When to start sending (UTC)")


# ─── Recipient schemas ───────────────────────────────────────────────


class RecipientRead(BaseModel):
    """Campaign recipient response."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    campaign_id: uuid.UUID
    contact_id: uuid.UUID
    status: RecipientStatus
    whatsapp_message_id: str | None
    sent_at: datetime | None
    delivered_at: datetime | None
    read_at: datetime | None
    error_message: str | None
    created_at: datetime


class RecipientList(BaseModel):
    """Paginated list of campaign recipients."""

    items: list[RecipientRead]
    total: int
    page: int
    page_size: int


# ─── Stats schemas ───────────────────────────────────────────────────


class CampaignStats(BaseModel):
    """Aggregated campaign delivery statistics."""

    total: int
    sent: int
    delivered: int
    read: int
    failed: int
    pending: int
    delivery_rate: float | None = None
    read_rate: float | None = None


class AudiencePreview(BaseModel):
    """Preview of targeted audience before sending."""

    count: int = Field(..., description="Total contacts matching the filter")
    sample: list[dict] = Field(
        default_factory=list,
        description="Sample of up to 5 matching contacts",
    )

File: frontend/src/types/campaign.ts
// ---------------------------------------------------------------------------
// Campaign / Publipostage types — mirrors backend schemas (campaign.py + enums.py)
// ---------------------------------------------------------------------------

export type CampaignStatus =
  | "draft"
  | "scheduled"
  | "sending"
  | "paused"
  | "completed"
  | "failed";

export type RecipientStatus =
  | "pending"
  | "sent"
  | "delivered"
  | "read"
  | "failed";

/** Inline stats stored in Campaign.stats JSONB column. */
export interface CampaignStatsInline {
  sent: number;
  delivered: number;
  read: number;
  failed: number;
  total: number;
}

/** Full campaign — mirrors CampaignRead from the backend. */
export interface Campaign {
  id: string;
  name: string;
  description: string | null;
  template_id: string;
  template_name: string;
  template_language: string;
  audience_filter: Record<string, unknown>;
  audience_count: number;
  variable_mapping: Record<string, string>;
  status: CampaignStatus;
  scheduled_at: string | null;
  started_at: string | null;
  completed_at: string | null;
  stats: CampaignStatsInline;
  created_by: string;
  created_at: string;
  updated_at: string;
}

/** Aggregated stats from GET /campaigns/{id}/stats. */
export interface CampaignStatsData {
  total: number;
  sent: number;
  delivered: number;
  read: number;
  failed: number;
  pending: number;
  delivery_rate: number | null;
  read_rate: number | null;
}

/** Single recipient from GET /campaigns/{id}/recipients. */
export interface CampaignRecipient {
  id: string;
  campaign_id: string;
  contact_id: string;
  status: RecipientStatus;
  whatsapp_message_id: string | null;
  sent_at: string | null;
  delivered_at: string | null;
  read_at: string | null;
  error_message: string | null;
  created_at: string;
}

/** Audience preview from POST /campaigns/{id}/preview. */
export interface AudiencePreview {
  count: number;
  sample: Array<Record<string, unknown>>;
}

/** Quota status from GET /campaigns/quota. */
export interface QuotaStatus {
  allowed: boolean;
  used: number;
  limit: number;
  remaining: number;
  percentage: number;
}

// ---------------------------------------------------------------------------
// Payloads
// ---------------------------------------------------------------------------

export interface CampaignCreatePayload {
  name: string;
  description?: string;
  template_id: string;
  template_name: string;
  template_language: string;
  audience_filter: Record<string, unknown>;
  variable_mapping?: Record<string, string>;
}

export interface CampaignUpdatePayload {
  name?: string;
  description?: string;
  audience_filter?: Record<string, unknown>;
  variable_mapping?: Record<string, string>;
}

export interface CampaignSchedulePayload {
  scheduled_at: string;
}

// ---------------------------------------------------------------------------
// Wizard form data (4-step wizard state)
// ---------------------------------------------------------------------------

export interface CampaignWizardData {
  // Step 1 — Template
  template_id: string;
  template_name: string;
  template_language: string;
  // Step 2 — Audience
  audience_tags: string[];
  audience_language: string;
  // Step 3 — Variables
  variable_mapping: Record<string, string>;
  // Step 4 — Schedule
  name: string;
  description: string;
  send_mode: "immediate" | "scheduled";
  scheduled_at: string;
}

/** WhatsApp template definition for the wizard. */
export interface WhatsAppTemplate {
  id: string;
  name: string;
  language: string;
  body: string;
  header?: string;
  footer?: string;
  buttons?: string[];
  variables: string[];
}
